### DISCUSSION TITLE
NotNullWhenAttribute for Tuples?

### DISCUSSION THREAD

**Discussion by Pentadome** [dotnet/csharplang#3116] [Q&A]
I have a method called 
`TryGetCustomAttribute<TAttribute>(MemberInfo member)`
That returns the Tuple: `(bool Success, TAttribute? Attribute)`
How do i indicate that Attribute is not null when Success is true? 

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 17290c4d-73f0-5ae1-e8ba-c039fe9fe9ca
* Version Independent ID: 59e0e3bb-b2e5-3ced-72d9-42e641b3897f
* Content: [Upgrade APIs for nullable reference types with attributes that define expectations for null values](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-attributes)
* Content Source: [docs/csharp/nullable-attributes.md](https://github.com/dotnet/docs/blob/master/docs/csharp/nullable-attributes.md)
* Product: **dotnet-csharp**
* Technology: **csharp-null-safety**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

**Comment by cartermp** [comment]
I don't think it's possible to apply this attribute to elements of a tuple type. Might be good to file this as a suggestion under dotnet/csharplang.

**Comment by BillWagner** [comment]
Thanks @cartermp I've transferred the issue to the csharplang repo.

**Comment by canton7** [comment]
I'm not arguing for or against this request, but I will just point out what's currently available.

Firstly, note that [this](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.customattributeextensions.getcustomattribute?view=netframework-4.8#System_Reflection_CustomAttributeExtensions_GetCustomAttribute__1_System_Reflection_MemberInfo_) already exists. This means you can call `member.GetCustomAttribute<SomeAttribute>()`, and it'll give you back the `SomeAttribute` if found, or `null` if it wasn't.

Secondly, the normal pattern for `TryXXX` methods is:

```cs
public static bool TryGetCustomAttribute<T>(MemberInfo member, [MaybeNullWhen(false)] out T attribute);
```

From personal experience, I've found this is easier to use than returning a `(bool success, T? value)`: for starters, you can use it in a single-line `if`:

```cs
if (TryGetCustomAttribute<SomeAttribute>(member, out var attribute))
{
   // use attribute
}
```

**Reply by chucker** [reply]
>Secondly, the normal pattern for TryXXX methods is:
>[..]
>From personal experience, I've found this is easier to use than returning a (bool success, T? value):

That's definitely 1) the correct pattern where possible, and 2) easier to use, but it doesn't work for `async`.

The closest approximation for `async` I can think of is indeed a tuple, and for that, we could indeed use a "not `null` when tuple field A is `true`".

**Comment by DavidArno** [comment]
@canton7

> From personal experience, I've found this is easier to use than returning a (bool success, T? value): for starters, you can use it in a single-line if...

You can with tuples too:
```cs
if (TryGetCustomAttribute<SomeAttribute>(member) is (true, {} attribute))
{
   // use attribute
}
```
Some people like `out` parameters. For those that don't, there really is no need to use them any more as tuples and pattern matching serve the same purpose.

And because that `{} attribute` tests that the  value isn't null and assigns it to a non-nullable variable, there's arguably no real need a `NotNullWhenAttribute` attribute.
